src/Analyser/NodeScopeResolver.php

```diff
@@ -950,6 +949,7 @@ private function processStmtNode(
 
 			do {
 				$prevScope = $bodyScope;
+				$bodyScope = $bodyScope->mergeWith($scope);
```

Comment:
this fixes the first failure, see also https://github.com/phpstan/phpstan-src/pull/1849
at least here removing it causes a test failure. but as mentioned there - I did not fully understand it tbh, it was do-while specific though

Reply:
Yeah, I guess it makes sense, the same thing is done (`$bodyScope = $bodyScope->mergeWith($scope);`) before the final analysis on line 978.